### ISSUE TITLE
The current Backoffice user shouldn't be able to reset his own password! (Security issue)

### ISSUE THREAD

**Issue by OmarElSergany** [umbraco/Umbraco-CMS#13942]
### Which Umbraco version are you using? (Please write the *exact* version, example: 10.1.0)

11.2.0

### Bug summary

When login to the backend (Backoffice) the current user can reset any other user password and change his own.
But he shouldn’t be able to reset his own.

### Specifics

Project : Umbraco.Web.BackOffice
Folder : Security
File : PasswordChanger.cs
Method : ChangePasswordWithIdentityAsync
 
There is if condition 
 
```
// Are we just changing another user/member's password?
if (passwordModel.OldPassword.IsNullOrWhiteSpace())
```
 
it’s assume that if the old password is null we are reset other users password.
 
But it can be used to reset the current user password! And if the user is the only admin attacker will get access and prevent the admin from login to the backoffice.

### Steps to reproduce

1. login to the Backoffice. ( or hijack open session).
2. click the user avatar.
3. click change password.
4. open any network traffic capturing software or http debugger (ex : Telerik Fiddler) and run it.
5. type any password in new and old password.
6. now back to  Fiddler and open composer tab and drag and drop the change password request from the requests list to edit it.
7.  remove the old password from request body.
 
Change : `{"newPassword":"1111333333","oldPassword":"1111"}`
To : `{"newPassword":"1111333333"}`
 
8. click execute to reset the password!

### Expected result / actual result

It reset the password instead of validating there is no old password and return Attempt.Fail.

**Suggested fix**
 
Just check if the passwordModel.Id is the current user id
 
```
            // Are we just changing another user/member's password?
            if (passwordModel.OldPassword.IsNullOrWhiteSpace())
            {
 
               

>                 if (userId == identityUser.Id)
>                 {
>                     return Attempt.Fail(new PasswordChangedModel
>                     {
>                         ChangeError = new ValidationResult("Password could not be verified", new[] { "oldPassword" })
>                     });
>                 }

 
 
                // ok, we should be able to reset it
                var resetToken = await userMgr.GeneratePasswordResetTokenAsync(identityUser);
 
                IdentityResult resetResult =
                    await userMgr.ChangePasswordWithResetAsync(userId, resetToken, passwordModel.NewPassword);
 
                if (resetResult.Succeeded == false)
                {
                    var errors = resetResult.Errors.ToErrorMessage();
                    _logger.LogWarning("Could not reset user password {PasswordErrors}", errors);
                    return Attempt.Fail(new PasswordChangedModel
                    {
                        ChangeError = new ValidationResult(errors, new[] { "value" })
                    });
                }
 
                return Attempt.Succeed(new PasswordChangedModel());
            }
```
 
Thank you

---
_This item has been added to our backlog [AB#27521](https://dev.azure.com/umbraco/243e7927-03b2-44e2-908f-d4ac7ea5daaa/_workitems/edit/27521)_

**Comment by kjac** [comment]
Hi @OmarElSergany,

Thank you for raising this and for digging into the code too 👏 

While we don't consider this critical, we do recognise that it should be fixed. We'll prioritise this in an upcoming sprint.

For good measure (and future reference once we pick this up), the proposed fix won't work, as it will always yield `true` no matter the context; `userId` will always be the same as `identityUser.Id` because we use `userId` to find `identityUser`:

![image](https://user-images.githubusercontent.com/7405322/224262300-3152c842-c0fe-4463-bbe8-bec2f56d61ba.png)

Instead we need to pass the current user from the executing context (i.e. the services or controllers invoking `ChangePasswordWithIdentityAsync`). This will likely end up as a breaking change to `IPasswordChanger` - if not binary breaking then at least functionally breaking.

**Comment by nikolajlauridsen** [comment]
Thanks again for reporting, this has been fixed in #14189 🎉
